File: packages/render/src/native/road-decals.ts
import type { NativeTextureMemberSource } from './texture-profile.js';

/**
 * The map's road decal derivative (`derived/road-decals/manifest.json`,
 * schema `simforge.map-road-decals.v1`, built by
 * `@simforge-oss/map-pipeline` `buildRoadDecals`, docs/engineering/map-road-decals.md):
 * the RoadRunner wear-decal layers of the master and their calibrated
 * opacity. The service applies it when the scene loads.
 */
export const NATIVE_ROAD_DECALS_MANIFEST = 'derived/road-decals/manifest.json';
const ROAD_DECALS_SCHEMA = 'simforge.map-road-decals.v1';

export interface NativeRoadDecalsPlan {
  readonly members: readonly string[];
  readonly manifestSha256: string;
  readonly buildKey: string;
  readonly opacityScale: number;
  readonly materials: number;
}

/**
 * `undefined` when the map carries no derivative (decals composite at their
 * authored opacity; the evidence says so). A derivative bound to another
 * master or malformed is an error, never a silent authored-opacity render.
 */
export async function planNativeRoadDecals(source: NativeTextureMemberSource): Promise<NativeRoadDecalsPlan | undefined> {
  const manifestSha256 = source.sha256(NATIVE_ROAD_DECALS_MANIFEST);
  if (!manifestSha256) return undefined;
  let manifest: { schema?: unknown; buildKey?: unknown; opacityScale?: unknown; materials?: unknown; source?: { master?: { path?: unknown; sha256?: unknown } } };
  try {
    manifest = JSON.parse(await source.readText(NATIVE_ROAD_DECALS_MANIFEST)) as typeof manifest;
  } catch (error) {
    throw new Error(`native_road_decals_invalid: ${NATIVE_ROAD_DECALS_MANIFEST} is not JSON (${(error as Error).message})`);
  }
  if (manifest.schema !== ROAD_DECALS_SCHEMA) throw new Error(`native_road_decals_invalid: schema ${String(manifest.schema)} (${ROAD_DECALS_SCHEMA})`);
  if (typeof manifest.buildKey !== 'string') throw new Error('native_road_decals_invalid: no buildKey');
  if (typeof manifest.opacityScale !== 'number' || !(manifest.opacityScale >= 0 && manifest.opacityScale <= 1)) {
    throw new Error('native_road_decals_invalid: opacityScale');
  }
  if (!Array.isArray(manifest.materials)) throw new Error('native_road_decals_invalid: materials');
  const master = manifest.source?.master;
  if (master?.path !== 'master.gltf' || master.sha256 !== source.sha256('master.gltf')) {
    throw new Error('native_road_decals_master_mismatch: the derivative was built from another master.gltf');
  }
  return {
    members: [NATIVE_ROAD_DECALS_MANIFEST],
    manifestSha256,
    buildKey: manifest.buildKey,
    opacityScale: manifest.opacityScale,
    materials: manifest.materials.length,
  };
}

File: native/crates/simforge-cli/src/render/road_decals.rs
//! The map's road decal derivative (`road-decals.ts`;
//! `derived/road-decals/manifest.json`, schema `simforge.map-road-decals.v1`):
//! the RoadRunner wear-decal layers and their calibrated opacity.

use serde_json::Value;

use super::error::{PlanError, PlanResult};
use super::map_closure::{MemberSource, MASTER_PATH};

pub const MANIFEST: &str = "derived/road-decals/manifest.json";
const SCHEMA: &str = "simforge.map-road-decals.v1";

#[derive(Debug, Clone, PartialEq, serde::Serialize)]
#[serde(rename_all = "camelCase")]
pub struct RoadDecalsPlan {
    pub members: Vec<String>,
    pub manifest_sha256: String,
    pub build_key: String,
    pub opacity_scale: f64,
    pub materials: usize,
}

/// `planNativeRoadDecals`: `None` when the map carries no derivative (decals
/// composite at authored opacity; the evidence records it). A derivative
/// bound to another master or malformed is an error.
pub fn plan_road_decals(source: &dyn MemberSource) -> PlanResult<Option<RoadDecalsPlan>> {
    let Some(manifest_sha256) = source.sha256(MANIFEST).map(str::to_owned) else {
        return Ok(None);
    };
    let manifest: Value = serde_json::from_str(&source.read_text(MANIFEST)?).map_err(|e| {
        PlanError::new(
            "native_road_decals_invalid",
            format!("{MANIFEST} is not JSON ({e})"),
        )
    })?;
    if manifest.get("schema").and_then(Value::as_str) != Some(SCHEMA) {
        let found = match manifest.get("schema") {
            None => "undefined".to_owned(),
            Some(Value::String(s)) => s.clone(),
            Some(other) => other.to_string(),
        };
        return Err(PlanError::new(
            "native_road_decals_invalid",
            format!("schema {found} ({SCHEMA})"),
        ));
    }
    let Some(build_key) = manifest.get("buildKey").and_then(Value::as_str) else {
        return Err(PlanError::new("native_road_decals_invalid", "no buildKey"));
    };
    let opacity_scale = manifest
        .get("opacityScale")
        .and_then(Value::as_f64)
        .filter(|v| (0.0..=1.0).contains(v));
    let Some(opacity_scale) = opacity_scale else {
        return Err(PlanError::new("native_road_decals_invalid", "opacityScale"));
    };
    let Some(materials) = manifest.get("materials").and_then(Value::as_array) else {
        return Err(PlanError::new("native_road_decals_invalid", "materials"));
    };
    let master = manifest.pointer("/source/master");
    let master_sha = master.and_then(|m| m.get("sha256")).and_then(Value::as_str);
    if master.and_then(|m| m.get("path")).and_then(Value::as_str) != Some(MASTER_PATH)
        || master_sha.is_none()
        || master_sha != source.sha256(MASTER_PATH)
    {
        return Err(PlanError::new(
            "native_road_decals_master_mismatch",
            "the derivative was built from another master.gltf",
        ));
    }
    Ok(Some(RoadDecalsPlan {
        members: vec![MANIFEST.to_owned()],
        manifest_sha256,
        build_key: build_key.to_owned(),
        opacity_scale,
        materials: materials.len(),
    }))
}

#[cfg(test)]
mod tests {
    //! Ported from `road-decals.test.ts`.
    use super::*;
    use crate::render::testing::FakeSource;
    use serde_json::json;

    fn source(manifest: Option<&Value>, master: &str) -> FakeSource {
        let s = FakeSource::default().file("master.gltf", master, "{}");
        match manifest {
            Some(m) => s.file(MANIFEST, &"c".repeat(64), &m.to_string()),
            None => s,
        }
    }

    fn manifest() -> Value {
        json!({
            "schema": SCHEMA, "buildKey": "b".repeat(64), "opacityScale": 0.2,
            "source": { "master": { "path": "master.gltf", "sha256": "a".repeat(64) } },
            "materials": [{ "index": 3, "name": "OilPath01_Road_Roads_Road_Layer1" }],
        })
    }

    #[test]
    fn is_absent_when_the_map_carries_none() {
        assert_eq!(
            plan_road_decals(&source(None, &"a".repeat(64))).unwrap(),
            None
        );
    }

    #[test]
    fn plans_the_manifest_member_and_reports_what_it_applies() {
        assert_eq!(
            plan_road_decals(&source(Some(&manifest()), &"a".repeat(64))).unwrap(),
            Some(RoadDecalsPlan {
                members: vec![MANIFEST.to_owned()],
                manifest_sha256: "c".repeat(64),
                build_key: "b".repeat(64),
                opacity_scale: 0.2,
                materials: 1,
            })
        );
    }

    #[test]
    fn refuses_a_derivative_from_another_master_never_renders_authored_opacity_silently() {
        assert_eq!(
            plan_road_decals(&source(Some(&manifest()), &"d".repeat(64)))
                .unwrap_err()
                .code,
            "native_road_decals_master_mismatch"
        );
        let mut bad = manifest();
        bad["opacityScale"] = json!(3);
        assert!(plan_road_decals(&source(Some(&bad), &"a".repeat(64)))
            .unwrap_err()
            .message
            .contains("opacityScale"));
    }
}
